From 9537db7ddf069c536443ad4341e451e93316ca9b Mon Sep 17 00:00:00 2001
From: Loc Vu <loc.vu.zn@renesas.com>
Date: Tue, 14 Jul 2020 14:59:38 +0700
Subject: [PATCH 038/433] ARM: shmobile: r8a7745: Add basic SoC support

- Add config for R8A7745 in Kconfig.debug.
- Add setup and sgx-quirk for R8A7745.
- Add more index macro for clock sources of R8A7745.

Signed-off-by: Loc Vu <loc.vu.zn@renesas.com>
Signed-off-by: Hao Bui <hao.bui.yg@renesas.com>
---
 arch/arm/Kconfig.debug                     |  4 +-
 arch/arm/mach-shmobile/Makefile            |  1 +
 arch/arm/mach-shmobile/setup-r8a7745.c     | 33 +++++++++++++++
 arch/arm/mach-shmobile/sgx-r8a7745-quirk.c | 68 ++++++++++++++++++++++++++++++
 include/dt-bindings/clock/r8a7745-clock.h  |  5 +++
 5 files changed, 109 insertions(+), 2 deletions(-)
 create mode 100644 arch/arm/mach-shmobile/setup-r8a7745.c
 create mode 100644 arch/arm/mach-shmobile/sgx-r8a7745-quirk.c

diff --git a/arch/arm/Kconfig.debug b/arch/arm/Kconfig.debug
index 7df0767..a8d4753 100644
--- a/arch/arm/Kconfig.debug
+++ b/arch/arm/Kconfig.debug
@@ -834,10 +834,10 @@ choice
 
 	config DEBUG_RCAR_GEN2_SCIF2
 		bool "Kernel low-level debugging messages via SCIF2 on R8A7794"
-		depends on ARCH_R8A7794
+		depends on ARCH_R8A7794 || ARCH_R8A7745
 		help
 		  Say Y here if you want kernel low-level debugging support
-		  via SCIF2 on Renesas R-Car E2 (R8A7794).
+		  via SCIF2 on Renesas R-Car E2 (R8A7794) or R8A7745.
 
 	config DEBUG_RCAR_GEN2_SCIF4
 		bool "Kernel low-level debugging messages via SCIF4 on R8A7745"
diff --git a/arch/arm/mach-shmobile/Makefile b/arch/arm/mach-shmobile/Makefile
index 13a4b38..a783af3 100644
--- a/arch/arm/mach-shmobile/Makefile
+++ b/arch/arm/mach-shmobile/Makefile
@@ -13,6 +13,7 @@ obj-$(CONFIG_ARCH_R8A7778)	+= setup-r8a7778.o
 obj-$(CONFIG_ARCH_R8A7779)	+= setup-r8a7779.o pm-r8a7779.o
 obj-$(CONFIG_ARCH_R8A7790)	+= setup-r8a7790.o
 obj-$(CONFIG_ARCH_R8A7791)	+= setup-r8a7791.o
+obj-$(CONFIG_ARCH_R8A7745)	+= setup-r8a7745.o sgx-r8a7745-quirk.o
 obj-$(CONFIG_ARCH_EMEV2)	+= setup-emev2.o
 obj-$(CONFIG_ARCH_R7S72100)	+= setup-r7s72100.o
 
diff --git a/arch/arm/mach-shmobile/setup-r8a7745.c b/arch/arm/mach-shmobile/setup-r8a7745.c
new file mode 100644
index 0000000..eea4f42
--- /dev/null
+++ b/arch/arm/mach-shmobile/setup-r8a7745.c
@@ -0,0 +1,33 @@
+/*
+ * r8a7745 processor support
+ *
+ * Copyright (C) 2014  Renesas Electronics Corporation
+ * Copyright (C) 2014  Ulrich Hecht
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 2 of the License.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#include <linux/of_platform.h>
+#include "common.h"
+#include "rcar-gen2.h"
+#include <asm/mach/arch.h>
+
+static const char * const r8a7745_boards_compat_dt[] __initconst = {
+	"renesas,r8a7745",
+	NULL,
+};
+
+DT_MACHINE_START(R8A7745_DT, "Generic R8A7745 (Flattened Device Tree)")
+	.init_early	= shmobile_init_delay,
+	.init_late	= shmobile_init_late,
+	.init_time	= rcar_gen2_timer_init,
+	.reserve	= rcar_gen2_reserve,
+	.dt_compat	= r8a7745_boards_compat_dt,
+MACHINE_END
diff --git a/arch/arm/mach-shmobile/sgx-r8a7745-quirk.c b/arch/arm/mach-shmobile/sgx-r8a7745-quirk.c
new file mode 100644
index 0000000..72e4f82
--- /dev/null
+++ b/arch/arm/mach-shmobile/sgx-r8a7745-quirk.c
@@ -0,0 +1,68 @@
+#include <linux/device.h>
+#include <linux/init.h>
+#include <linux/io.h>
+#include <linux/notifier.h>
+#include <linux/of.h>
+#include <linux/platform_device.h>
+#include <linux/clk.h>
+
+#define DEV_NAME	"fd800000.sgx"
+
+static struct clk *clk;
+
+static int r8a7745_sgx_notifier_call(struct notifier_block *nb,
+					unsigned long event, void *data)
+{
+	struct device *dev = data;
+
+	switch (event) {
+	case BUS_NOTIFY_BIND_DRIVER:
+		if (!strcmp(DEV_NAME, dev_name(dev))) {
+
+			if (IS_ERR(clk))
+				break;
+			if (clk_prepare_enable(clk) < 0)
+				pr_warn("SGX Quirk can not enable clk\n");
+			pr_info("SGX quirk clk is enabled\n");
+		}
+		break;
+
+	case BUS_NOTIFY_UNBOUND_DRIVER:
+		if (!strcmp(DEV_NAME, dev_name(dev))) {
+			if (IS_ERR(clk))
+				break;
+			clk_disable_unprepare(clk);
+			pr_info("SGX quirk clk is disabled\n");
+		}
+		break;
+	}
+	return NOTIFY_DONE;
+}
+
+static struct notifier_block platform_nb = {
+	.notifier_call = r8a7745_sgx_notifier_call,
+};
+
+static int __init sgx_r8a7745_quirk_init(void)
+{
+	static int once;
+
+	if (once++)
+		return -ENOMEM;
+
+	if (!of_machine_is_compatible("renesas,r8a7745"))
+		return -ENODEV;
+
+	clk = clk_get(NULL, "vcp0");
+	if (IS_ERR(clk)) {
+		pr_warn("SGX Quirk can not get clk\n");
+		return -ENOENT;
+	}
+
+	bus_register_notifier(&platform_bus_type, &platform_nb);
+
+	pr_info("Installing sgx clock quirk\n");
+	return 0;
+}
+
+arch_initcall(sgx_r8a7745_quirk_init);
diff --git a/include/dt-bindings/clock/r8a7745-clock.h b/include/dt-bindings/clock/r8a7745-clock.h
index f1659d5..5c8bcfa 100644
--- a/include/dt-bindings/clock/r8a7745-clock.h
+++ b/include/dt-bindings/clock/r8a7745-clock.h
@@ -72,6 +72,7 @@
 
 /* MSTP5 */
 #define R8A7745_CLK_AUDIO_DMAC0		2
+#define R8A7745_CLK_THERMAL		22
 #define R8A7745_CLK_PWM			23
 
 /* MSTP7 */
@@ -132,12 +133,16 @@
 #define R8A7745_CLK_SCU_DVC0		19
 #define R8A7745_CLK_SCU_CTU1_MIX1	20
 #define R8A7745_CLK_SCU_CTU0_MIX0	21
+#define R8A7745_CLK_SCU_SRC9		22
+#define R8A7745_CLK_SCU_SRC8		23
+#define R8A7745_CLK_SCU_SRC7		24
 #define R8A7745_CLK_SCU_SRC6		25
 #define R8A7745_CLK_SCU_SRC5		26
 #define R8A7745_CLK_SCU_SRC4		27
 #define R8A7745_CLK_SCU_SRC3		28
 #define R8A7745_CLK_SCU_SRC2		29
 #define R8A7745_CLK_SCU_SRC1		30
+#define R8A7745_CLK_SCU_SRC0		31
 
 /* MSTP11 */
 #define R8A7745_CLK_SCIFA3		6
-- 
2.7.4

